### ISSUE TITLE
Server-side entity transform data streaming and client-side data interpolation

### ISSUE THREAD

**Issue by Irrelon** [Irrelon/ige#6]
# Server-Side

Add component for data streaming to allow clients to receive updates from the server about changes to an entity's transform data just as 1.0.0 already does but with more control over which clients receive the updates.

System should have two modes:
- Auto-mode: Use a callback model that will ask a method to return true or false for each client as to whether the update should be sent to them or not. If no callback is defined, send to all clients each tick whenever new data is available. 
- Advanced-mode: Allow the programmer to control when and how data is synced with which clients. Provide a method in the IgeEntity class called streamSync() that will take a client id (or array of ids) as a parameter and send the transform update to those clients.
- Transform data should be generated only once each tick on request (not automatically, only if required to conserve CPU) and subsequent requests for the data in the same tick should return the cache. Similarly, if no transform data has changed between ticks, cached data should be returned.
- Write a component that allows a "visibility radius". The radius will determine when that entity should be sent from the server to the client (and also when any transform updates should be streamed). If the client's player entity moves inside the radius, the entity data is then streamed to the client. When the player entity moves outside of the radius the entity streaming stops for that client and the entity is destroyed client-side. This allows mechanics like hidden objects to be created that only become available in client-side memory (so it cannot be examined to determine location) once the player is in "range" of the entity.
# Client-Side

On the client-side, once transform update stream data has been received it needs to be applied via interpolation. The system in 1.0.0 is already perfect for porting to 1.1.0. Some updates would make this a much more advanced system:

When data for an entity is first received by a client it needs to know:
- The classId of the entity (what class the entity is created with)
- The ID of the entity
- The transform data

Everything else should be static or at least, is not part of the data stream. This system therefore relies on the programmer having correctly set up their classes for use in a client/server fashion. Take this example:
- We create a new class that extends IgeEntity called "MyPlayer"
- MyPlayer can be instantiated on both client and server
- In the MyPlayer init() method the programmer determines what properties and default values are set. This can be conditional based on if the class is being instantiated on client or server using the ige.isServer boolean flag
- Things like a texture only need to be applied on the client-side since the server will not deal with rendering
- The idea is that when the client receives a message from the server saying "Create this entity with a classId of MyPlayer and assign it this ID", everything the client needs to know in order to set that entity up correctly should already be in the client init() or other extended methods. The server should not have to send ANYTHING to the client beyond the classId (MyPlayer), ID to assign this new entity and it's transform data (this does not include dimensions... they should also be set during class init())

When the client receives a data stream for an entity that does not yet exist, it can simply do this:

var newEntity = new classId().id(newEntityId);

The classId will actually be a string so we will look up our ACTUAL class object from the global class store, something like:

var newEntity = new ige.classStore(classId)().id(newEntityId);

Comments, questions and ideas related to this post are welcome.
